Lib/test/test_pathlib.py

```diff
@@ -2478,6 +2478,226 @@ def test_complex_symlinks_relative(self):
     def test_complex_symlinks_relative_dot_dot(self):
         self._check_complex_symlinks(os.path.join('dirA', '..'))
 
+class WalkTests(unittest.TestCase):
+    """Tests for Path.walk()
+
+    They are mostly just copies of os.walk tests converted to use Paths
+    because os.walk tests are already mature and cover many edge cases
+    that we could miss writing new tests.
+
+    It is not combined with Path* tests because its setup is aimed at
+    testing a bit more complex cases than Path's setup.
+    """
+
+    # Wrapper to hide minor differences between different implementations
+    # of Path.walk (for example, Path.fwalk)
+    def walk(self, top, **kwargs):
+        return top.walk(**kwargs)
+
+    def setUp(self):
+        P = pathlib.Path
+        join = os.path.join
+        self.addCleanup(os_helper.rmtree, os_helper.TESTFN)
+
+        # Build:
+        #     TESTFN/
+        #       TEST1/              a file kid and two directory kids
+        #         tmp1
+        #         SUB1/             a file kid and a directory kid
+        #           tmp2
+        #           SUB11/          no kids
+        #         SUB2/             a file kid and a dirsymlink kid
+        #           tmp3
+        #           SUB21/          not readable
+        #             tmp5
+        #           link/           a symlink to TEST2
+        #           broken_link
+        #           broken_link2
+        #           broken_link3
+        #       TEST2/
+        #         tmp4              a lone file
+        self.walk_path = P(os_helper.TESTFN, "TEST1")
+        self.sub1_path = self.walk_path / "SUB1"
+        self.sub11_path = self.sub1_path / "SUB11"
+        self.sub2_path = self.walk_path / "SUB2"
+        sub21_path= self.sub2_path / "SUB21"
+        tmp1_path = self.walk_path / "tmp1"
+        tmp2_path = self.sub1_path / "tmp2"
+        tmp3_path = self.sub2_path / "tmp3"
+        tmp5_path = sub21_path / "tmp3"
+        self.link_path = self.sub2_path / "link"
+        t2_path = P(os_helper.TESTFN, "TEST2")
+        tmp4_path = P(os_helper.TESTFN, "TEST2", "tmp4")
+        broken_link_path = self.sub2_path / "broken_link"
+        broken_link2_path = self.sub2_path / "broken_link2"
+        broken_link3_path = self.sub2_path / "broken_link3"
+
+        # Create stuff.
+        os.makedirs(self.sub11_path)
+        os.makedirs(self.sub2_path)
+        os.makedirs(sub21_path)
+        os.makedirs(t2_path)
+
+        for path in tmp1_path, tmp2_path, tmp3_path, tmp4_path, tmp5_path:
+            with open(path, "x", encoding='utf-8') as f:
+                f.write(f"I'm {path} and proud of it.  Blame test_os.\n")
+
+        if os_helper.can_symlink():
+            os.symlink(os.path.abspath(t2_path), self.link_path)
+            os.symlink('broken', broken_link_path, True)
+            os.symlink(join('tmp3', 'broken'), broken_link2_path, True)
+            os.symlink(join('SUB21', 'tmp5'), broken_link3_path, True)
```

Comment:
```suggestion
            os.symlink(P('tmp3', 'broken'), broken_link2_path, True)
            os.symlink(P('SUB21', 'tmp5'), broken_link3_path, True)
```

(and remove the `join = os.path.join` bit above)